File: linkedListIntersection.js
// https://leetcode.com/problems/intersection-of-two-linked-lists/

const getIntersectionNode = function(one, two) {
	let oneLength = 0;
	let twoLength = 0;
	let o = one;
	let t = two;
	while (o.next || t.next) {
		if (o.next) {
			o = o.next;
			oneLength++;
		}
		if (t.next) {
			t = t.next;
			twoLength++;
		}
	}
	if (o !== t) return null;
	let diff = oneLength - twoLength;
	o = one;
	t = two;
	if (diff > 0) {
		for (let i = 0; i < diff; i++) {
			o = o.next;
		}
	} else if (diff < 0) {
		for (let i = 0; i < Math.abs(diff); i++) {
			t = t.next;
		}
	}
	while (true) {
		if (o === t) {
			return o;
		}
		o = o.next;
		t = t.next;
	}
};

File: linkedListIntersection.py
# https://leetcode.com/problems/intersection-of-two-linked-lists/

class ListNode:
    def __init__(self, val, nex = None):
        self.val = val
        self.next = nex

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not(headA and headB):
            return None
        size1 = 0
        p1 = headA
        size2 = 0
        p2 = headB
        while p1.next:
            size1 += 1
            p1 = p1.next
        while p2.next:
            size2 += 1
            p2 = p2.next
        p1 = headA
        p2 = headB
        while size1 > size2:
            p1 = p1.next
            size1 -= 1
        while size2 > size1:
            p2 = p2.next
            size2 -= 1
        while p1 and p2:
            if p1 is p2:
                return p1
            p1 = p1.next
            p2 = p2.next
        return None

test = Solution()
interList = ListNode(2, ListNode(4))
l1 = ListNode(3, interList)
l2 = ListNode(1, ListNode(9, ListNode(1 , interList)))
print(test.getIntersectionNode(l1, l2).next.val)

